// nativescript-plugin-firebase/mlkit/facedetection/index.ios.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var image_source_1 = require("tns-core-modules/image-source");
var facedetection_common_1 = require("./facedetection-common");
var MLKitFaceDetection = /** @class */ (function (_super) {
    __extends(MLKitFaceDetection, _super);
    function MLKitFaceDetection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MLKitFaceDetection.prototype.createDetector = function () {
        return getDetector();
    };
    MLKitFaceDetection.prototype.createSuccessListener = function () {
        var _this = this;
        return function (faces, error) {
            if (error !== null) {
                console.log(error.localizedDescription);
            }
            else if (faces !== null && faces.count > 0) {
                var result = {
                    faces: []
                };
                for (var i = 0, l = faces.count; i < l; i++) {
                    var face = faces.objectAtIndex(i);
                    console.log(">> face: " + face);
                    result.faces.push({
                        smilingProbability: face.hasSmilingProbability ? face.smilingProbability : undefined,
                        leftEyeOpenProbability: face.hasLeftEyeOpenProbability ? face.leftEyeOpenProbability : undefined,
                        rightEyeOpenProbability: face.hasRightEyeOpenProbability ? face.rightEyeOpenProbability : undefined
                    });
                }
                console.log(">>> notify " + MLKitFaceDetection.scanResultEvent + " with " + JSON.stringify(result.faces));
                _this.notify({
                    eventName: MLKitFaceDetection.scanResultEvent,
                    object: _this,
                    value: result
                });
            }
        };
    };
    MLKitFaceDetection.prototype.rotateRecording = function () {
        return false;
    };
    return MLKitFaceDetection;
}(facedetection_common_1.MLKitFaceDetection));
exports.MLKitFaceDetection = MLKitFaceDetection;
function getDetector() {
    // TODO make configurable (see #704)
    var firVision = FIRVision.vision();
    var options = FIRVisionFaceDetectorOptions.new();
    options.modeType = 2 /* Accurate */;
    options.landmarkType = 2 /* All */;
    options.classificationType = 2 /* All */;
    options.minFaceSize = 0.1;
    // options.isTrackingEnabled = true;
    return firVision.faceDetectorWithOptions(options);
}
// TODO somehow this function doesn't work.. probably because of the passed image, but I can't find the cause.. the live camera version works great tho
function detectFacesOnDevice(options) {
    return new Promise(function (resolve, reject) {
        try {
            var faceDetector = getDetector();
            faceDetector.detectInImageCompletion(getImage(options), function (faces, error) {
                if (error !== null) {
                    reject(error.localizedDescription);
                }
                else if (faces !== null) {
                    var result = {
                        faces: []
                    };
                    for (var i = 0, l = faces.count; i < l; i++) {
                        var face = faces.objectAtIndex(i);
                        result.faces.push({
                            smilingProbability: face.hasSmilingProbability ? face.smilingProbability : undefined,
                            leftEyeOpenProbability: face.hasLeftEyeOpenProbability ? face.leftEyeOpenProbability : undefined,
                            rightEyeOpenProbability: face.hasRightEyeOpenProbability ? face.rightEyeOpenProbability : undefined
                        });
                    }
                    resolve(result);
                }
            });
        }
        catch (ex) {
            console.log("Error in firebase.mlkit.detectFaces: " + ex);
            reject(ex);
        }
    });
}
exports.detectFacesOnDevice = detectFacesOnDevice;
// TODO resize the image (here), like the example app?
function getImage(options) {
    var image = options.image instanceof image_source_1.ImageSource ? options.image.ios : options.image.imageSource.ios;
    console.log(">> image.imageOrientation: " + image.imageOrientation);
    var fIRVisionImageMetadata = FIRVisionImageMetadata.new();
    // fIRVisionImageMetadata.orientation = FIRVisionDetectorImageOrientation.TopLeft;
    // const randomOrientation = Math.floor(Math.random() * 8) + 1;
    // console.log(">>> randomOrientation: " + randomOrientation);
    fIRVisionImageMetadata.orientation = 1;
    var fIRVisionImage = FIRVisionImage.alloc().initWithImage(image);
    fIRVisionImage.metadata = fIRVisionImageMetadata;
    return fIRVisionImage;
}
